### PR TITLE
Full experimental iavl implementation

### PR DESCRIPTION
**Pull request by yihuang**
implemented set/delete/get functions, the root hashes are checked against the golang version.
Don't affect the existing cli functionalities.

### CODE REVIEW

File: iavl/iavl.py [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+        self.update_height_size(ndb)
+        new.update_height_size(ndb)
+        return new
+
+    def update_height_size(self, ndb: NodeDB):
+        lnode = self.left_node(ndb)
+        rnode = self.right_node(ndb)
+        self.height = max(lnode.height, rnode.height) + 1
+        self.size = lnode.size + rnode.size
+
```
Comment (mmsqe) [inline iavl/iavl.py:185]: maybe can align [calcBalance](https://github.com/cosmos/iavl/blob/aff636602ca8dc07a333fae893d49e163d7a7e43/node.go#L508)
```suggestion
    def calc_balance(self, ndb: NodeDB):
        return self.left_node(ndb).height - self.right_node(ndb).height
```

File: iavl/iavl.py [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+    def balance(self, ndb: NodeDB, version: int):
+        lnode = self.left_node(ndb)
+        rnode = self.right_node(ndb)
+        balance = lnode.height - rnode.height
+        if balance > 1:
+            lbalance = lnode.left_node(ndb).height - lnode.right_node(ndb).height
```
Comment (mmsqe) [inline iavl/iavl.py:191]:
```suggestion
            lbalance = lnode.calc_balance(ndb)
```

File: iavl/iavl.py [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+            if lbalance >= 0:
+                # left left
+                return self.rotate_right(ndb, version)
+            else:
+                # left right
+                self.left_node_ref = lnode.rotate_left(ndb, version)
+                return self.rotate_right(ndb, version)
+        elif balance < -1:
+            rbalance = rnode.left_node(ndb).height - rnode.right_node(ndb).height
```
Comment (mmsqe) [inline iavl/iavl.py:200]:
```suggestion
            rbalance = rnode.calc_balance(ndb)
```

File: iavl/iavl.py (hunk 2 above)
Comment (yihuang) [reply iavl/iavl.py:191]: I've considered that, the reason I didn't make it a separate function is that, a) the lnode and rnode are used later in this function, b) `calc_balance` is only used here anyway, it won't be reused anywhere else.

Comment (mmsqe) [reply iavl/iavl.py:191]: make sense, but I'll make it reuse if > 2 places needed

File: iavl/iavl.py [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+    def update_height_size(self, ndb: NodeDB):
+        lnode = self.left_node(ndb)
+        rnode = self.right_node(ndb)
+        self.height = max(lnode.height, rnode.height) + 1
+        self.size = lnode.size + rnode.size
+
+    def balance(self, ndb: NodeDB, version: int):
+        lnode = self.left_node(ndb)
+        rnode = self.right_node(ndb)
+        balance = lnode.height - rnode.height
```
Comment (mmsqe) [inline iavl/iavl.py:189]:
```suggestion
        balance = self.calc_balance(ndb)
```

File: iavl/iavl.py [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+            return None, None
+    else:
+        turn_left = key < node.key
+        if turn_left:
+            value, new_child = remove_recursive(ndb, key, node.left_node_ref, version)
+        else:
+            value, new_child = remove_recursive(ndb, key, node.right_node_ref, version)
+
+        if value is None:
+            return
```
[the remark this answers is not in the document]
Comment (yihuang) [reply iavl/iavl.py:302]: good catch, will need a new test case to cover as well, should look at the test coverage as well

File: iavl/iavl.py (hunk 1 above)
Comment (yihuang) [reply iavl/iavl.py:186]: done

File: iavl/iavl.py (hunk 2 above)
Comment (yihuang) [reply iavl/iavl.py]: done

File: iavl/iavl.py (hunk 3 above)
Comment (yihuang) [reply iavl/iavl.py]: done

File: iavl/iavl.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (yihuang) [reply iavl/iavl.py]: fixed, reached 100% test coverage.

File: iavl/iavl.py (hunk 4 above)
Comment (yihuang) [reply iavl/iavl.py]: done

**Review by mmsqe — APPROVED** [review]
so will we handle orphan later?

**Comment by yihuang** [comment]
> so will we handle orphan later?

yeah, I guess better in a separate PR.

**Pull request merged**